tokio/src/macros/scoped_tls.rs

```diff
@@ -0,0 +1,80 @@
+use crate::loom::thread::LocalKey;
+
+use std::cell::Cell;
+use std::marker;
+
+/// The macro. See the module level documentation for the description and examples.
+#[macro_export]
+macro_rules! scoped_thread_local {
+    ($(#[$attrs:meta])* $vis:vis static $name:ident: $ty:ty) => (
+        $(#[$attrs])*
+        $vis static $name: $crate::macros::scoped_tls::ScopedKey<$ty>
+            = $crate::macros::scoped_tls::ScopedKey {
+                inner: {
+                    thread_local!(static FOO: ::std::cell::Cell<usize> = {
+                        ::std::cell::Cell::new(0)
+                    });
+                    &FOO
+                },
+                _marker: ::std::marker::PhantomData,
+            };
+    )
+}
+
+/// Type representing a thread local storage key corresponding to a reference
+/// to the type parameter `T`.
+pub(crate) struct ScopedKey<T> {
+    #[doc(hidden)]
+    pub(crate) inner: &'static LocalKey<Cell<usize>>,
```

Comment:
Why is this a `usize` instead of a `*mut T`?

Reply:
Work around lifetime issues.

Reply:
Of `T` specifically? Otherwise they should be pretty much the same, except that `*mut T` doesn't erase the fact that it's a pointer. Alternatively `*mut ()` (or I guess `NonNull<()>`).